nav2_collision_monitor/src/source_base.cpp

```diff
@@ -0,0 +1,192 @@
+  const std::string base_frame_id,
+  const double transform_tolerance,
+  const double max_time_shift)
+: node_(node), tf_buffer_(tf_buffer), source_type_(SOURCE_BASE), source_name_(source_name),
+  source_topic_(""), source_frame_id_(""), base_frame_id_(base_frame_id),
+  transform_tolerance_(transform_tolerance), max_time_shift_(max_time_shift)
+{
+  auto node_sptr = node_.lock();
+  if (node_sptr) {
+    RCLCPP_INFO(node_sptr->get_logger(), "Creating SourceBase");
+  }
+}
+
+SourceBase::~SourceBase()
+{
+  auto node = node_.lock();
+  if (node) {
+    RCLCPP_INFO(node->get_logger(), "Destroying SourceBase");
+  }
+
+  tf_buffer_.reset();
+}
+
+bool SourceBase::getParameters()
+{
+  auto node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  try {
+    nav2_util::declare_parameter_if_not_declared(
+      node, source_name_ + ".topic",
+      rclcpp::ParameterValue("scan"));  // Set deafult topic for laser scanner
+    source_topic_ = node->get_parameter(source_name_ + ".topic").as_string();
+  } catch (const std::exception & ex) {
+    RCLCPP_ERROR(
+      node->get_logger(), "Error while getting basic source parameters: %s", ex.what());
+    return false;
+  }
+
+  return true;
+}
+
+SourceType SourceBase::getSourceType()
+{
+  return source_type_;
+}
+
+std::string SourceBase::getSourceTypeStr()
+{
+  if (source_type_ == SOURCE_BASE) {
+    return "base";
+  } else if (source_type_ == SCAN) {
+    return "scan";
+  } else {  // source_type_ == POINTCLOUD
+    return "pointcloud";
+  }
+}
+
+bool SourceBase::getTransform(
+  const std::string to_frame,
+  const std::string from_frame,
+  tf2::Transform & tf2_transform)
+{
+  auto node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  if (!tf_buffer_) {
+    RCLCPP_ERROR(node->get_logger(), "tf_buffer_ is not initialized");
+    return false;
+  }
+
+  geometry_msgs::msg::TransformStamped transform;
+  tf2_transform.setIdentity();  // initialize by identical transform
+
+  if (from_frame == to_frame) {
+    // tf2_transform is already identical. Do nothing.
+    return true;
+  }
+
+  try {
+    const tf2::Duration tr_tol = tf2::durationFromSec(transform_tolerance_);
+    // Obtaining the transform to get data from one to another frame
+    transform = tf_buffer_->lookupTransform(
+      to_frame, from_frame, tf2::TimePointZero, tr_tol);
+  } catch (tf2::TransformException & e) {
+    RCLCPP_ERROR(
+      node->get_logger(),
+      "Failed to get \"%s\"->\"%s\" frame transform: %s",
+      source_frame_id_.c_str(), base_frame_id_.c_str(), e.what());
+    return false;
+  }
+
+  // Convert TransformStamped to TF2 transform
+  tf2::fromMsg(transform.transform, tf2_transform);
+  return true;
+}
+
+void SourceBase::getData(
+  std::vector<Point> & data, const rclcpp::Time & curr_time, const Velocity & velocity)
+{
+  std::lock_guard<mutex_t> lock(data_mutex_);
+
+  fixData(curr_time, velocity);
```

Comment:
What is "fix"? This is a strange word, is something broken? :laughing:

Reply:
In latest model, I've switched to `SourceBase::getSourceBaseTransform()` routine with time `lookupTransform()` interpolation usage for data sources. For Scan example: https://github.com/ros-planning/navigation2/pull/2982/files#diff-a8ab77e548af62fe331c646d4df405f23654b0a652d6d63718a4c0c1f6ee1ab7R71 and https://github.com/ros-planning/navigation2/pull/2982/files#diff-a8ab77e548af62fe331c646d4df405f23654b0a652d6d63718a4c0c1f6ee1ab7R84